### DISCUSSION TITLE
Escaping characters for strings in the Tasmota commands

### DISCUSSION THREAD

**Discussion by gkarag** [arendst/Tasmota#14353] [Q&A] [answered]
Hello community,

I had two similar problems recently for which I couldn't find the right spell:
- Assigning Tasmota var in the console to a string starting with a space or containing spaces only
- Put `[` and `]` on a display using the `DisplayText` command

In both cases I could not use the characters directly and I couldn't find the proper escape sequence.

I am using the latest official version 10.1.0 on ESP32. Any help on this subject?

**Comment by sfromis** [comment, marked as answer]
Docs for [Displays](https://tasmota.github.io/docs/Displays/#fonts) has tilde as escape character:
> The ~ character followed by a hex byte can define any character code.

Assuming that the blanks you want are also for display purposes, I'd expect escaped blanks to work for this.

**Reply by gkarag** [reply]
Your assumption is correct. Thanks for pointing me to the right place in the documentation.

**Reply by gkarag** [reply]
Unfortunately using variable with escape sequences for display purposes does not work from within a non-berry rule. For example if I want to set a variable so it would display as `[]` I set it to `~5B~5D`.  Then if I create a rule that has `DisplayText %var1%`, it shows as `~5B~5D` on the display. If I read the variable from a berry script and invoke from there `tasmota.cmd('DisplayText ~5B~5D')` it shows `[]` on the display as I expect. I tested it with the latest development release and it had the same behavior.
Have I missed something?

**Reply by sfromis** [reply]
```
00:53:48.082 CMD: rule1 on event#test do DisplayText %var1% endon
00:53:57.664 CMD: var1 ~5B~5D
00:54:03.932 CMD: event test
00:54:03.967 RUL: EVENT#TEST performs "DisplayText ~5B~5D"
```
And I see [] on a SSD1306 display. No problems with escape via a classic rule. Tested on both ESP8266 and ESP32, dev binary.

Anyway, with ESP32 I'd just skip classic rules most of the time, with the greatly enhanced features with Berry. Of course, classic rules should still work, as it does when I test.

**Reply by gkarag** [reply]
Strange, your example worked. Here is the rule  I used and it still misbehaves when `var2` contains `~` sequences. The sequence `~F8C` at the end works as expected:
```
ON DS18B20#TEMPERATURE DO DisplayText [zy0f0s1] %var2%  %var3%   %var1% [y10f0s2] %value% ~F8C ENDON
```
I have tried rule like the one below and it displayed `[]` again
```
rule1 on event#test do DisplayText [zy0f0s1] %var1% endon
```
For the record I am using ESP32 with with a 0.96 inch OLED display that has no other identifiers than Module 12864

Edit: For the record I am using a classical rule because I had some concurrency issues with the berry rules and besides the problem with the escaping it looks that it works fine.

Anyway I resorted to a solution with other special characters stored directly that doesn't use `~` escaping in the variables. Thanks for your explanations!

**Reply by sfromis** [reply]
Looks like it has nothing to do with variables, but some quirk in the display driver parsing where it may not always honor escape sequences.

In general it is a good plan to pare examples down to the simplest possible, like I did. Instead of making assumptions about what triggered the issue, better verify how little it takes to make the difference.

This should have shown how it is not a question of variables, rules or Berry (what you did there was very different), as you can get the same issue doing the same DisplayText from the command line, with values included directly (instead of variables).
